replica: remove rwlock for protecting iteration over storage group map

raphaelsc opened the pull request:
rwlock was added to protect iterations against concurrent updates to the map.

the updates can happen when allocating a new tablet replica or removing an old one (tablet cleanup).

the rwlock is very problematic because it can result in topology changes blocked, as updating token metadata takes the exclusive lock, which is serialized with table wide ops like split / major / explicit flush (and those can take a long time).

to get rid of the lock, we can copy the storage group map and guard individual groups with a gate (not a problem since map is expected to have a maximum of ~100 elements). so cleanup can close that gate (carefully closed after stopping individual groups such that migrations aren't blocked by long-running ops like major), and ongoing iterations (e.g. triggered by nodetool flush) can skip a group that was closed, as such a group is being migrated out.

Fixes #18821.

```
WRITE
=====

./build/release/scylla perf-simple-query --smp 1 --memory 2G --initial-tablets 10 --tablets --write

- BEFORE

65559.52 tps ( 59.6 allocs/op,  16.4 logallocs/op,  14.3 tasks/op,   52841 insns/op,   30946 cycles/op,        0 errors)
67408.05 tps ( 59.3 allocs/op,  16.0 logallocs/op,  14.3 tasks/op,   53018 insns/op,   30874 cycles/op,        0 errors)
67714.72 tps ( 59.3 allocs/op,  16.0 logallocs/op,  14.3 tasks/op,   53026 insns/op,   30881 cycles/op,        0 errors)
67825.57 tps ( 59.3 allocs/op,  16.0 logallocs/op,  14.3 tasks/op,   53015 insns/op,   30821 cycles/op,        0 errors)
67810.74 tps ( 59.3 allocs/op,  16.0 logallocs/op,  14.3 tasks/op,   53009 insns/op,   30828 cycles/op,        0 errors)

         throughput: mean=67263.72 standard-deviation=967.40 median=67714.72 median-absolute-deviation=547.02 maximum=67825.57 minimum=65559.52
instructions_per_op: mean=52981.61 standard-deviation=79.09 median=53014.96 median-absolute-deviation=36.54 maximum=53025.79 minimum=52840.56
  cpu_cycles_per_op: mean=30869.90 standard-deviation=50.23 median=30874.06 median-absolute-deviation=42.11 maximum=30945.94 minimum=30820.89


- AFTER
65448.76 tps ( 59.5 allocs/op,  16.4 logallocs/op,  14.3 tasks/op,   52788 insns/op,   31013 cycles/op,        0 errors)
67290.83 tps ( 59.3 allocs/op,  16.0 logallocs/op,  14.3 tasks/op,   53025 insns/op,   30950 cycles/op,        0 errors)
67646.81 tps ( 59.3 allocs/op,  16.0 logallocs/op,  14.3 tasks/op,   53025 insns/op,   30909 cycles/op,        0 errors)
67565.90 tps ( 59.3 allocs/op,  16.0 logallocs/op,  14.3 tasks/op,   53058 insns/op,   30951 cycles/op,        0 errors)
67537.32 tps ( 59.3 allocs/op,  16.0 logallocs/op,  14.3 tasks/op,   52983 insns/op,   30963 cycles/op,        0 errors)

         throughput: mean=67097.93 standard-deviation=931.44 median=67537.32 median-absolute-deviation=467.97 maximum=67646.81 minimum=65448.76
instructions_per_op: mean=52975.85 standard-deviation=108.07 median=53024.55 median-absolute-deviation=49.45 maximum=53057.99 minimum=52788.49
  cpu_cycles_per_op: mean=30957.17 standard-deviation=37.43 median=30951.31 median-absolute-deviation=7.51 maximum=31013.01 minimum=30908.62


READ
=====

./build/release/scylla perf-simple-query --smp 1 --memory 2G --initial-tablets 10 --tablets

- BEFORE

79423.36 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.2 tasks/op,   41840 insns/op,   26820 cycles/op,        0 errors)
81076.70 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.2 tasks/op,   41837 insns/op,   26583 cycles/op,        0 errors)
80927.36 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.2 tasks/op,   41829 insns/op,   26629 cycles/op,        0 errors)
80539.44 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.2 tasks/op,   41841 insns/op,   26735 cycles/op,        0 errors)
80793.10 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.2 tasks/op,   41864 insns/op,   26662 cycles/op,        0 errors)

         throughput: mean=80551.99 standard-deviation=661.12 median=80793.10 median-absolute-deviation=375.37 maximum=81076.70 minimum=79423.36
instructions_per_op: mean=41842.20 standard-deviation=13.26 median=41840.14 median-absolute-deviation=5.68 maximum=41864.50 minimum=41829.29
  cpu_cycles_per_op: mean=26685.88 standard-deviation=93.31 median=26662.18 median-absolute-deviation=56.47 maximum=26820.08 minimum=26582.68


- AFTER
79464.70 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.2 tasks/op,   41799 insns/op,   26761 cycles/op,        0 errors)
80954.58 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.2 tasks/op,   41803 insns/op,   26605 cycles/op,        0 errors)
81160.90 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.2 tasks/op,   41811 insns/op,   26555 cycles/op,        0 errors)
81263.10 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.2 tasks/op,   41814 insns/op,   26527 cycles/op,        0 errors)
81162.97 tps ( 63.1 allocs/op,   0.0 logallocs/op,  14.2 tasks/op,   41806 insns/op,   26549 cycles/op,        0 errors)

         throughput: mean=80801.25 standard-deviation=755.54 median=81160.90 median-absolute-deviation=361.72 maximum=81263.10 minimum=79464.70
instructions_per_op: mean=41806.47 standard-deviation=5.85 median=41806.05 median-absolute-deviation=4.05 maximum=41813.86 minimum=41799.36
  cpu_cycles_per_op: mean=26599.22 standard-deviation=94.84 median=26554.54 median-absolute-deviation=50.51 maximum=26761.06 minimum=26527.05
```

bhalevy commented on replica/table.cc:
```diff
@@ -550,9 +550,27 @@ void table::enable_off_strategy_trigger() {
+        auto holder = sg->async_gate().hold();
+        co_await f(*sg);
+    });
+    co_return;
```
nit: co_return isn't required

bhalevy commented on replica/table.cc:
```diff
@@ -550,9 +550,27 @@ void table::enable_off_strategy_trigger() {
 storage_group_manager::~storage_group_manager() = default;
 
+future<> storage_group_manager::parallel_foreach_storage_group(std::function<future<>(storage_group&)> f) {
+    auto storage_groups = _storage_groups;
```
This isn't required since coroutine::parallel_for_each does not preempt when invoking the function on each element (same for legacy parallel_for_each).
It may yield only when co_awaiting the resulting futures.

That said, it's not wrong either.

raphaelsc replied:
you're right, but it's better to be explicit here right?

bhalevy replied:
It's a short list and just a copy of lw_shared_ptr so it shouldn't be terrible.
Does it make sense to use boost::copy_ramge to copy the shared group ptrs into a linear vector rather than copying the flat hash map?

avikivity replied:
Doesn't the caller care that its callback was executed for a storage group that was removed?

raphaelsc replied:
> Doesn't the caller care that its callback was executed for a storage group that was removed?

I don't think so. such procedures are used for table wide ops (e.g. nodetool flush), and if we have a tablet being migrated out, it's ok that this table wide op skips a storage group that was removed.

e.g.

1) tablet is being migrated out
2) runs a major compact / flush
3) loop takes a copy of storage map
4) cleanup removes storage for the tablet
5) loop reaches the storage group for that tablet and skips it because it no longer belongs to this node

do you think we need to enforce it somehow other than documentation?

raphaelsc replied:
it's a trade off I think we're willing to make, to not block migrations / topology changes by taking a lock in cleanup, or update_effective_replication_map() (which keeps old erm alive till completion).

avikivity replied:
I think it's okay.
 - nodetools-style operations are inherently racy wrt migration
 - more global operations should use cluster-wide methods to prevent the race

raphaelsc replied:
I add doc to compaction_group.hh explaining the approach and scenarios to be handled.

raphaelsc replied:
added*

bhalevy commented on replica/table.cc:
```diff
@@ -550,9 +550,27 @@ void table::enable_off_strategy_trigger() {
 future<> storage_group_manager::for_each_storage_group_gently(std::function<future<>(size_t, storage_group&)> f) {
-    rwlock::holder shared_lock = co_await get_rwlock().hold_read_lock();
-    for (auto& [id, sg]: _storage_groups) {
+    auto storage_groups = _storage_groups;
```
here's the copy is justified since we may yield in the loop body

Deexie replied:
If we copy storage_groups map, preempt and the map will be modified, `storage_group_manager::_storage_groups` and `storage_group_manager::_compaction_groups` may become inconsistent: 
a storage group is removed, while compaction group is still there (`storage_group_manager::_compaction_groups` is an intrusive list with auto-unlink).

raphaelsc replied:
We need to remove this intrusive list. Probably a good opportunity

raphaelsc replied:
done

bhalevy commented on replica/table.cc:
```diff
@@ -3551,7 +3575,6 @@ future<> table::cleanup_compaction_groups(database& db, db::system_keyspace& sys
 future<> table::cleanup_tablet(database& db, db::system_keyspace& sys_ks, locator::tablet_id tid) {
     auto holder = async_gate().hold();
-    rwlock::holder exclusive_lock = co_await _sg_manager->get_rwlock().hold_write_lock();
     auto& sg = storage_group_for_id(tid.value());
```
don't we need to enter the storage_group gate here?

bhalevy replied:
Otherwise, we should copy the sg compaction_groups  synchronously and pass that to `clear_inactive_reads_for_tablet` (we could do it there too, but I think it's cleaner to do it this way since the scope of usage is clearer)

raphaelsc replied:
> Otherwise, we should copy the sg compaction_groups synchronously and pass that to `clear_inactive_reads_for_tablet` (we could do it there too, but I think it's cleaner to do it this way since the scope of usage is clearer)

I think it's not a problem, since cleanup_tablet() is the one deallocating the group (happens in the end).

bhalevy commented on replica/table.cc:
```diff
@@ -3563,8 +3586,6 @@ future<> table::cleanup_tablet(database& db, db::system_keyspace& sys_ks, locato
     auto holder = async_gate().hold();
-    // Hold shared lock to keep storage group alive.
-    rwlock::holder shared_lock = co_await _sg_manager->get_rwlock().hold_read_lock();
     auto& sg = storage_group_for_id(tid.value());
```
ditto, shouldn't we enter the storage_group gate?

raphaelsc commented on replica/table.cc:
```diff
@@ -3515,11 +3530,20 @@ future<> table::clear_inactive_reads_for_tablet(database& db, storage_group& sg)
     }
+    // carefully closed after stopping individual groups, such that cleanup won't be blocked on ongoing compaction.
+    co_await _async_gate.close();
+}
```
@bhalevy there's something I realized I don't handle yet. what if tablet is migrated out during major? 

let's say tablet X is migrated out while major is running on it (with storage_group::async_gate taken). but given that we're closing gate after stopping groups (because we don't want tablet cleanup to wait on compaction), it can happen major fail due to it being stopped on compaction_group::stop(), which results in table wide op failing entirely.

raphaelsc replied:
I don't think it's a problem since compaction_stopped_exception (triggered by compaction_group::stop()) is swallowed by compaction manager (used in shutdown path).

bhalevy replied:
right.  regular compaction could also run when a tablet is migrated and we should stop it too.

Deexie commented on replica/table.cc:
```diff
@@ -3515,11 +3530,20 @@ future<> table::clear_inactive_reads_for_tablet(database& db, storage_group& sg)
+    }
     // Synchronizes with in-flight writes if any, and also takes care of flushing if needed.
-    for (auto& cg_ptr : sg.compaction_groups()) {
+    for (auto& cg_ptr : compaction_groups()) {
```
A compaction group can be stopped while `table::parallel_foreach_compaction_group` still uses it

raphaelsc replied:
This is similar to shutdown and already handled with stop(), which closes gates, stops ongoing compaction and waits for flush if any. I also made compaction_group a shared_ptr to avoid lifetime issues.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9951/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9951/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2258/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10279/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9951/)
#### Build Details:
- Duration: 2 hr 32 min
- Builder: spider6.cloudius-systems.com

raphaelsc commented:
v2:
- gets rid of fragile intrusive list (another source of truth which can result in inconsistency).
- added documentation to explain scenarios and how each is handled without rwlock
- improved storage_group::stop() by launching close of gate first, and waiting for it after stopping compaction groups (to avoid waiting on ongoing compactions)

Deexie commented on replica/table.cc:
```diff
@@ -582,17 +589,19 @@ storage_group& storage_group_manager::storage_group_for_id(const schema_ptr& s,
 
     compaction_group& get_compaction_group() const noexcept {
-        return const_cast<compaction_group&>(*_compaction_groups.begin());
+        return *_single_cg;
```
Why not `*_single_sg->main_compaction_group().get()`?

raphaelsc replied:
sure, why not. even though this is a hot path, compiler should be able to optimize it easily

raphaelsc commented:
v3:
- fixed scylla-gdb.py
- added another test to increase confidence
- hardened it further by holding guard of compaction groups in parallel_foreach_compaction_group

raphaelsc commented:
@Deexie please take another look.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10035/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10035/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_experimental_raft/test_tablets
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10035/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2317/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10336/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10035/)
#### Build Details:
- Duration: 6 hr 37 min
- Builder: spider4.cloudius-systems.com

bhalevy commented on replica/table.cc:
```diff
@@ -368,8 +368,8 @@ future<std::vector<locked_cell>> table::lock_counter_cells(const mutation& m, db
 std::vector<memtable*> table::active_memtables() {
-    return boost::copy_range<std::vector<memtable*>>(compaction_groups() | boost::adaptors::transformed([] (compaction_group& cg) {
-        return &cg.memtables()->active_memtable();
+    return boost::copy_range<std::vector<memtable*>>(compaction_groups() | boost::adaptors::transformed([] (compaction_group* cg) {
```
This adds an allocation, in `compaction_groups()`.
We better get the size first to reserve capacity for the returned vector, then iterate through all storage/compaction groups, rather than returning a temporary vector to iterate over.

Please run `perf-simple-query --write` to compare performance / allocations impact of this change.

bhalevy replied:
I see that the function is currently not called on the hot read/write path, but it might be in the future.
But there is no reason to over-simplify it, leaving a time bomb behind.

raphaelsc replied:
Yes, not a hot path today and I am planning to get rid of compaction_groups() in a near-future patch. replace it by for_each_active_memtable() is a good call, I think.

bhalevy commented on replica/table.cc:
```diff
@@ -974,14 +978,38 @@ compaction_group& table::compaction_group_for_sstable(const sstables::shared_sst
-compaction_group_list& table::compaction_groups() const noexcept {
-    return _sg_manager->compaction_groups();
+// FIXME: not used in hot paths, but can be removed later to avoid allocation (using chunked to reduce reallocations).
+utils::chunked_vector<compaction_group*> table::compaction_groups() const {
```
I think it would be better to introduce for_each_compaction_group helper rather than introducing a function + FIXME :)

bhalevy commented on replica/table.cc:
```diff
@@ -974,14 +978,38 @@ compaction_group& table::compaction_group_for_sstable(const sstables::shared_sst
 
-compaction_group_list& table::compaction_groups() const noexcept {
-    return _sg_manager->compaction_groups();
+// FIXME: not used in hot paths, but can be removed later to avoid allocation (using chunked to reduce reallocations).
```
Why don't let the caller use `for_each_compaction_group` instead?

bhalevy commented on replica/table.cc:
```diff
@@ -550,17 +550,28 @@ void table::enable_off_strategy_trigger() {
-    co_await coroutine::parallel_for_each(_storage_groups, [&] (std::pair<size_t, const std::unique_ptr<storage_group>&> p) -> future<> {
-        co_await f(p.first, *p.second.get());
+// exception-less attempt to hold gate.
+static std::optional<gate::holder> try_hold_gate(gate& g) noexcept {
```
Can be added to seastar as `gate::try_hold` as a follow up.

bhalevy replied:
https://github.com/scylladb/seastar/pull/2315

raphaelsc replied:
thanks.

bhalevy commented on replica/table.cc:
```diff
@@ -550,17 +550,28 @@ void table::enable_off_strategy_trigger() {
+
+future<> storage_group_manager::parallel_foreach_storage_group(std::function<future<>(storage_group&)> f) {
+    auto storage_groups = boost::copy_range<std::vector<storage_group_ptr>>(_storage_groups | boost::adaptors::map_values);
+    co_await coroutine::parallel_for_each(storage_groups, [&] (const storage_group_ptr& sg) -> future<> {
```
parallel_for_each doesn't yield when iterating the range so the copy above is superfluous.
The lambda can accept a `storage_group_ptr` by value instead to extend its lifetime (besides entering its gate)

raphaelsc replied:
I think we discussed this in previous patch and was ok with this extra copy here to make it explicit the intent, rather than relying on implementation detail of parallel_for_each(). table wide ops aren't part of hot path. do you agree?

raphaelsc replied:
You're right, it's fine as long as we make it an invariant.

bhalevy commented on replica/table.cc:
```diff
@@ -550,17 +550,28 @@ void table::enable_off_strategy_trigger() {
-        co_await f(id, *sg);
+future<> storage_group_manager::for_each_storage_group_gently(std::function<future<>(storage_group&)> f) {
+    auto storage_groups = boost::copy_range<std::vector<storage_group_ptr>>(_storage_groups | boost::adaptors::map_values);
+    for (auto& sg: storage_groups) {
```
Also, why did you change the interface and stopped passing the storage_group id to the called function?
Better do it in a separate patch if it required/desired and provide the motivation for it there.

raphaelsc replied:
I documented it. But nevermind, will extract it from this big patch. thanks

raphaelsc replied:
> Also, why did you change the interface and stopped passing the storage_group id to the called function? Better do it in a separate patch if it required/desired and provide the motivation for it there.

the motivation was only created within 2nd patch, so I decided to keep it there while adding explanation for motivation in the commit message.

bhalevy commented on replica/table.cc:
```diff
@@ -983,16 +997,26 @@ utils::chunked_vector<compaction_group*> table::compaction_groups() const {
+            tlogger.info("foreach_compaction_group_wait: released");
+        });
+
         for (auto& cg : sg.compaction_groups()) {
```
We lost the inter-compaction_group parallelism here (pre-existing issue).
Why aren't we calling parallel_for_each here?

raphaelsc replied:
we should. sometimes I am worried about queuing too much work, that cannot happen simulteneaously anyway. for example, only one split is allowed per shard. But we're also bound by number of tablets in a shard, so it's not like it can grow unboundly. Ok, will parallelize compaction groups of a single tablet.

bhalevy commented on replica/table.cc:
```diff
@@ -2039,14 +2063,15 @@ future<> compaction_group::stop() noexcept {
         co_return;
     }
-    co_await _async_gate.close();
+    auto closed_gate_fut = _async_gate.close();
```
better use `coroutine::as_future(_async_gate.close());`, similar to `flush_future` just below.

That said, why closing the gate in the background is safe?

bhalevy commented on replica/table.cc:
```diff
@@ -3545,11 +3566,24 @@ future<> table::clear_inactive_reads_for_tablet(database& db, storage_group& sg)
+    // Carefully waits for close of gate after stopping compaction groups, since we don't want
+    // to wait on an ongoing compaction, *but* start it earlier to prevent iterations from
+    // picking this group that is being stopped.
+    auto closed_gate_fut = _async_gate.close();
```
please use `coroutine::as_future`

raphaelsc replied:
ok.

bhalevy commented on scylla-gdb.py:
```diff
@@ -4505,8 +4508,11 @@ def invoke(self, arg, from_tty):
-                for cg in intrusive_list(sg_manager["_compaction_groups"], link='_list_hook'):
-                    scylla_memtables.dump_compaction_group_memtables(cg)
+                for (sg_id, sg_ptr) in absl_container(sg_manager["_storage_groups"]):
+                    sg = seastar_lw_shared_ptr(sg_ptr).get()
```
Doesn't this lose backward compatibility?

raphaelsc replied:
6.0 is already out, so it's indeed better to preserve compatibility.

bhalevy commented:
@raphaelsc we must get perf-simple-query results (both read and write) before and after this change
to make sure there is no regressions (or even better, show some progression ;-))

Deexie commented on replica/table.cc:
```diff
@@ -2149,14 +2174,14 @@ future<> tablet_storage_group_manager::handle_tablet_split_completion(const loca
             auto group_id = first_new_id + i;
             split_ready_groups[i]->update_id_and_range(group_id, new_tmap.get_token_range(locator::tablet_id(group_id)));
-            new_storage_groups[group_id] = std::make_unique<storage_group>(std::move(split_ready_groups[i]));
+            new_storage_groups[group_id] = make_lw_shared<storage_group>(std::move(split_ready_groups[i]));
```
Is it ok that a loop can touch compaction groups which have their id and range updated?

raphaelsc replied:
for example

let's say there are 10 tablets in this shard

there will be 30 compaction groups (left, main, and right groups for each tablet). left contains 1st half of range, main contains data spanning both halves, and right the 2nd half.

left, main and right share the same tablet id. showing we cannot rely on it for uniqueness.

all id and range tell you is which tablet id a group belongs to, and the token range that group spans, respectively.

I don't think a loop over groups need tablet id stability, since the loop is there for table-wide ops like major or flush. loops that need tablet id stability won't yield (see for_each_storage_group usage in tablet_sstable_set and split completion also must set all groups without yielding and that's what it does and must do always for many reasons).

And the range that we set during split completion will still be correct (before and after). we're just making it more fine-grained based on the new tablet count.

raphaelsc commented:
> @raphaelsc we must get perf-simple-query results (both read and write) before and after this change to make sure there is no regressions (or even better, show some progression ;-))

I wouldn't really expect any improvement, other than not stalling topology changes anymore, which is already a huge deal. I will be happy if first result shows there is no regression :-)

raphaelsc commented:
v4:
- get rid of table::compaction_groups()
- change table::active_memtables() to table::for_each_active_memtable().
- use coroutine::as_future() when closing gate
- parallelize iteration at compaction group level too
- not lose backward compatibility in scylla-gdb.py
- perf regression test included in cover letter

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10182/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10182/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/commitlog_test
:small_blue_diamond: boost/memtable_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10182/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10461/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2436/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10182/)
#### Build Details:
- Duration: 6 hr 14 min
- Builder: spider3.cloudius-systems.com

raphaelsc commented:
@bhalevy please review again.

bhalevy commented on replica/table.cc:
```diff
@@ -772,20 +786,34 @@ compaction_group_ptr& storage_group::select_compaction_group(locator::tablet_ran
+    return cgs;
+}
+
+utils::small_vector<const compaction_group*, 3> storage_group::compaction_groups() const {
```
nit: noexcept as well

bhalevy commented on replica/table.cc:
```diff
@@ -974,17 +1002,34 @@ compaction_group& table::compaction_group_for_sstable(const sstables::shared_sst
-    rwlock::holder shared_lock = co_await _sg_manager->get_rwlock().hold_read_lock();
-    co_await coroutine::parallel_for_each(compaction_groups(), [&] (compaction_group& cg) {
-        return action(cg);
+void table::for_each_const_compaction_group(std::function<void(const compaction_group&)> action) const {
```
nit: do we really need a different function name?
isn't the const qualifier and the `action` type enough to differentiate between them?

raphaelsc replied:
good question, the compiler should be able to tell the difference by looking at lambda param type. if you don't have any major comment, I am happy to send a patch as a followup.

bhalevy replied:
Seems to work. See https://godbolt.org/z/qM7P8xze9

bhalevy commented on replica/table.cc:
```diff
@@ -2153,14 +2177,18 @@ future<> compaction_group::stop() noexcept {
         co_await seastar::coroutine::return_exception_ptr(flush_future.get_exception());
     }
+    auto closed_gate_fut_res = co_await std::move(closed_gate_fut);
+    if (closed_gate_fut_res.failed()) {
```
keeping the as_future_awaiter in a variable is bit weird. my bad for suggesting it.
I take my words back. If we want `_async_gate.close()` to happen in the background, it's ok to do:
```
    auto closed_gate_fut = _async_gate.close();
...
    co_await std::move(closed_gate_fut);
```

we're in stop so stop() is never expected to fail, and if it did, it's ok to throw inside the coroutine.
we're not on the hot path anyways.

so let's simplify this.
again, I apologize for going back and forth on this

bhalevy commented on replica/table.cc:
```diff
@@ -3592,13 +3616,29 @@ future<> table::clear_inactive_reads_for_tablet(database& db, storage_group& sg)
+    // Carefully waits for close of gate after stopping compaction groups, since we don't want
+    // to wait on an ongoing compaction, *but* start it earlier to prevent iterations from
+    // picking this group that is being stopped.
+    auto closed_gate_fut = coroutine::as_future(_async_gate.close());
```
ditto

bhalevy reviewed (changes requested):
looks good overall.
some nitpicking, but most substantial comment is to revert the use of as_future
I changed my mind about it since we expect (and need) the original future to happen in the backtground.
Usually with as_future, one would do: `auto x = co_await coroutine::as_future(fut)` which is akin to `.then_wrapped` in the sense that the future as awaited for, so `x` is always ready, and may contain an exception, but this isn't suitable if we don't want to await `fut`.

raphaelsc commented:
v5:
- rebased
- annotated 'storage_group::compaction_groups() const' with noexcept
- went back to previous approach of closing gate in stop()

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10300/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10300/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/commitlog_test
:small_blue_diamond: boost/memtable_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10300/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10556/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2531/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10300/)
#### Failed Tests (1/39088):
- [test_concurrent_restart](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10300/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_concurrent_restart) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_concurrent_restart)
#### Build Details:
- Duration: 14 hr
- Builder: i-011c41b6a7e3d625a (m5ad.8xlarge)

bhalevy commented on replica/table.cc:
```diff
@@ -550,10 +555,28 @@ void table::enable_off_strategy_trigger() {
-    for (auto& [id, sg]: _storage_groups) {
-        co_await f(id, *sg);
+// exception-less attempt to hold gate.
+// TODO: move it to seastar.
```
Coming up in https://github.com/scylladb/seastar/commit/6acb4cfdd9624ae05f7564df3210955bf5dc96d1

raphaelsc commented:
> * [test_concurrent_restart](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10300/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_concurrent_restart) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_concurrent_restart)

that's known issue: https://github.com/scylladb/scylladb/issues/15192

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10340/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10340/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/commitlog_test
:small_blue_diamond: boost/memtable_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10340/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10573/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2551/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10340/)
#### Failed Tests (1/39333):
- [test_tablet_drain_failure_during_decommission.dev.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10340/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_dev_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission.dev.5)
#### Build Details:
- Duration: 6 hr 24 min
- Builder: spider6.cloudius-systems.com

raphaelsc commented:
> * test_tablet_drain_failure_during_decommission

opened https://github.com/scylladb/scylladb/issues/19707, failure is unrelated.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10363/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10363/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/commitlog_test
:small_blue_diamond: boost/memtable_test
:small_blue_diamond: boost/mutation_test
:small_blue_diamond: topology_experimental_raft/test_tablets
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10363/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10593/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2574/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10363/)
#### Build Details:
- Duration: 6 hr 28 min
- Builder: spider3.cloudius-systems.com

raphaelsc commented:
@denesb please merge
